File: packages/sdk-typescript/test/webhook-signature.test.ts
import assert from "node:assert/strict";
import { createHmac } from "node:crypto";
import { describe, it } from "node:test";
import { verifyWebhookSignature } from "../src/webhook-signature.js";

describe("verifyWebhookSignature", () => {
  const secret = "whsec_test_secret";
  const rawBody = JSON.stringify({ eventType: "sales.lead.created", payload: { id: "1" } });

  function sign(body: string, withSecret = secret): string {
    return `sha256=${createHmac("sha256", withSecret).update(body).digest("hex")}`;
  }

  it("accepts a correctly signed payload", () => {
    assert.equal(verifyWebhookSignature(secret, rawBody, sign(rawBody)), true);
  });

  it("rejects a signature computed with the wrong secret", () => {
    assert.equal(verifyWebhookSignature(secret, rawBody, sign(rawBody, "wrong-secret")), false);
  });

  it("rejects a tampered body", () => {
    const tampered = JSON.stringify({ eventType: "sales.lead.created", payload: { id: "2" } });
    assert.equal(verifyWebhookSignature(secret, tampered, sign(rawBody)), false);
  });

  it("rejects a malformed signature header", () => {
    assert.equal(verifyWebhookSignature(secret, rawBody, "not-a-signature"), false);
  });
});

File: packages/sdk-flutter/test/webhook_signature_test.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';
import 'package:test/test.dart';
import 'package:voltx_sdk/voltx_sdk.dart';

void main() {
  const secret = 'whsec_test_secret';
  final rawBody = jsonEncode({
    'eventType': 'sales.lead.created',
    'payload': {'id': '1'},
  });

  String sign(String body, {String withSecret = secret}) {
    final digest = Hmac(sha256, utf8.encode(withSecret)).convert(utf8.encode(body));
    return 'sha256=$digest';
  }

  test('accepts a correctly signed payload', () {
    expect(verifyWebhookSignature(secret, rawBody, sign(rawBody)), isTrue);
  });

  test('rejects a signature computed with the wrong secret', () {
    expect(verifyWebhookSignature(secret, rawBody, sign(rawBody, withSecret: 'wrong-secret')), isFalse);
  });

  test('rejects a tampered body', () {
    final tampered = jsonEncode({
      'eventType': 'sales.lead.created',
      'payload': {'id': '2'},
    });
    expect(verifyWebhookSignature(secret, tampered, sign(rawBody)), isFalse);
  });

  test('rejects a malformed signature header', () {
    expect(verifyWebhookSignature(secret, rawBody, 'not-a-signature'), isFalse);
  });
}
